### PR TITLE
feat(serverless-spark): Add serverless-spark source with list_batches tool

### PR DESCRIPTION
**Pull request by dborowitz**
## Description

Add a new source `serverless-spark` for connecting to [Google Cloud Serverless for Apache Spark](https://cloud.google.com/dataproc-serverless/docs), along with a single simple tool `serverless-spark-list-batches`.

One outstanding though possibly trivial question is how to address naming/branding/source splitting. The official product name is "Google Cloud Serverless for Apache Spark", though this is a relatively recent rebranding of "Dataproc Serverless". I figured for the source and tool names we could use the short version "serverless-spark", but I've tried to retain the officially correct name in documentation. Granted, it's quite a mouthful and sticks out compared to the other names in Toolbox. On the other hand, I don't want to imply that this works with any Spark hosting environment besides Google's.  I'll double-check the branding guidelines with our PMs, but I appreciate Toolbox maintainers' feedback as well.

## PR Checklist

> Thank you for opening a Pull Request! Before submitting your PR, there are a
> few things you can do to make sure it goes smoothly:

- [ ] Make sure you reviewed
  [CONTRIBUTING.md](https://github.com/googleapis/genai-toolbox/blob/main/CONTRIBUTING.md)
- [ ] Make sure to open an issue as a
  [bug/issue](https://github.com/googleapis/genai-toolbox/issues/new/choose)
  before writing your code! That way we can discuss the change, evaluate
  designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)
- [ ] Make sure to add `!` if this involve a breaking change

🛠️ Part of  #1689

### CODE REVIEW

File: tests/serverlessspark/serverless_spark_integration_test.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+	runListBatchesTest(t)
+}
+
+// runListBatchesTest invokes the running list-batches tool and ensures it returns the correct
+// number of results. It can run successfully against any GCP project that has at least 2 succeeded
+// or failed Serverless Spark batches, of any age.
+func runListBatchesTest(t *testing.T) {
+	requestBody := bytes.NewBuffer([]byte(`{"pageSize": 2, "filter": "state = SUCCEEDED OR state = FAILED"}`))
+	req, err := http.NewRequest(http.MethodPost, "http://127.0.0.1:5000/api/tool/list-batches/invoke", requestBody)
```
Comment (duwenxin99) [inline tests/serverlessspark/serverless_spark_integration_test.go]: Can we add a more comprehensive golang test table here to make sure all the edge cases are tested? Ideally we want to test for every parameter including the `pageToken`. We should also be testing edge cases like empty results etc. The generic authentication feature testing should also be included. Feel free to refer to the [BigQuery tests ](https://github.com/googleapis/genai-toolbox/blob/a1a427878934499ea45f7355f04e5e3911c2ec15/tests/bigquery/bigquery_integration_test.go#L1733)as examples. Thank you!

Comment (dborowitz) [reply tests/serverlessspark/serverless_spark_integration_test.go]: Done.

To keep it compatible with arbitrary test projects, I took the approach of listing results with the underlying API directly, then comparing a possibly-paginated version of the results from the tool with the underlying results.

Note this is inherently racy, but the hope is that we won't be frequently creating new batches in the integration test project. If it turns out to be an issue, we can always filter to batches created >30m ago or something.

**Comment by dborowitz** [comment]
I've converted to draft to ensure we don't merge before confirming the documentation/branding story with my PM. Other than that, I've resolved the one comment from @duwenxin99 about integration tests, and it's ready for review.

**Comment by dborowitz** [comment]
Updated the docs after consulting with PM, and added docs for the prebuilt toolset, which I had missed before.

**Comment by duwenxin99** [comment]
@dborowitz Thanks for the test update! Could you add auth test cases to every tool like BQ does [here](https://github.com/googleapis/genai-toolbox/blob/9a55b804821a6ccfcd157bcfaee7e599c4a5cb63/tests/bigquery/bigquery_integration_test.go#L300) to make sure the auth feature is working correctly. An `authSource` config  should be added like [this](https://github.com/googleapis/genai-toolbox/blob/9a55b804821a6ccfcd157bcfaee7e599c4a5cb63/tests/common.go#L37) and the [test invocation](https://github.com/googleapis/genai-toolbox/blob/9a55b804821a6ccfcd157bcfaee7e599c4a5cb63/tests/bigquery/bigquery_integration_test.go#L517) should include a token header. Thanks!

**Comment by dborowitz** [comment]
@duwenxin99 I'm not sure testing the `authSource` is necessary? My tool currently doesn't support client OAuth, and relies entirely on gcloud ADCs. I don't expect any tools to have authorized parameters.

Or are you saying that's client OAuth/authorized invocations are a requirement for any new tool implementation nowadays?

File: internal/tools/serverlessspark/serverlesssparklistbatches/serverlesssparklistbatches.go [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+	}
+	inputSchema, _ := allParameters.McpManifest()
+
+	mcpManifest := tools.McpManifest{
+		Name:        cfg.Name,
+		Description: desc,
+		InputSchema: inputSchema,
+	}
+
+	return Tool{
```
Comment (duwenxin99) [inline internal/tools/serverlessspark/serverlesssparklistbatches/serverlesssparklistbatches.go:92]: We need to pass the `AuthRequired` parameter into the Tool struct for the authorized invocation feature to work.

**Comment by dborowitz** [comment]
Chatted offline. I was confusing authorized invocations with the underlying auth used by the source. I don't particularly see a need for authorized invocations (e.g. to support a multi-tenant server setup), but it's basically free and we can at least test it works correctly.

(Last push was just to rebase on master, new tests are forthcoming.)

File: internal/tools/serverlessspark/serverlesssparklistbatches/serverlesssparklistbatches.go (hunk 2 above)
Comment (dborowitz) [reply internal/tools/serverlessspark/serverlesssparklistbatches/serverlesssparklistbatches.go:92]: Done

**Comment by dborowitz** [comment]
Updated `integration.cloudbuild.yaml` as requested over chat.

File: tests/serverlessspark/serverless_spark_integration_test.go [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+	"github.com/googleapis/genai-toolbox/internal/testutils"
+	"github.com/googleapis/genai-toolbox/internal/tools/serverlessspark/serverlesssparklistbatches"
+	"github.com/googleapis/genai-toolbox/tests"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+)
+
+var (
+	serverlessSparkProject  = os.Getenv("SERVERLESS_SPARK_PROJECT")
+	serverlessSparkLocation = os.Getenv("SERVERLESS_SPARK_LOCATION")
```
Comment (duwenxin99) [inline tests/serverlessspark/serverless_spark_integration_test.go:41]: This needs to match the env var name passed in from cloudbuild.yaml

File: .ci/integration.cloudbuild.yaml [hunk 4]
```diff
@@ -723,6 +723,25 @@ steps:
...
           oracle
+
+  - id: "serverless-spark"
+    name: golang:1
+    waitFor: ["compile-test-binary"]
+    entrypoint: /bin/bash
+    env:
+      - "GOPATH=/gopath"
+      - "SERVERLESS_SPARK_PROJECT=$PROJECT_ID"
+      - "SERVERLESS_SPARK_LOCATION=$_REGION"
```
Comment (duwenxin99) [inline .ci/integration.cloudbuild.yaml:734]: We need to specify a default in the `substitutions` section

File: tests/serverlessspark/serverless_spark_integration_test.go (hunk 3 above)
Comment (dborowitz) [reply tests/serverlessspark/serverless_spark_integration_test.go:41]: They do match, or am I missing something? This is in cloudbuild.yaml:

```
      - "SERVERLESS_SPARK_PROJECT=$PROJECT_ID"
      - "SERVERLESS_SPARK_LOCATION=$_REGION"
```

File: .ci/integration.cloudbuild.yaml (hunk 4 above)
Comment (dborowitz) [reply .ci/integration.cloudbuild.yaml:734]: Line 854 has:

```
  _REGION: "us-central1"
```

(which matches the region that I created batches in on the official test project)

What else needs to be specified?

Comment (duwenxin99) [reply .ci/integration.cloudbuild.yaml:734]: Oh that's great! I thought we needed a different region for the serverless spark but it's good that we can reuse it

**Review by duwenxin99 — APPROVED** [review]
